sshd-core/src/main/java/org/apache/sshd/common/channel/ChannelAsyncOutputStream.java

```diff
@@ -90,52 +93,174 @@ public byte getCommandType() {
+     *
+     * @throws WritePendingException if a concurrent write is attempted
+     */
     @Override
-    public synchronized IoWriteFuture writeBuffer(Buffer buffer) throws IOException {
+    public IoWriteFuture writeBuffer(Buffer buffer) throws IOException {
         if (isClosing()) {
-            throw new EOFException("Closing: " + state);
+            throw new EOFException("Closing: " + writeState);
         }
 
         IoWriteFutureImpl future = new IoWriteFutureImpl(packetWriteId, buffer);
-        if (!pendingWrite.compareAndSet(null, future)) {
-            throw new WritePendingException("A write operation is already pending");
+        synchronized (writeState) {
+            if (writeState.isClosing()) { // Double check.
+                throw new EOFException("Closing: " + writeState);
+            }
+            if (writeState.writeInProgress) {
+                throw new WritePendingException("A write operation is already pending");
+            }
+            writeState.lastWrite = future;
+            writeState.pendingWrite = future;
+            writeState.writeInProgress = true;
+            writeState.waitingOnIo = false;
         }
         doWriteIfPossible(false);
         return future;
     }
 
     @Override
     protected void preClose() {
-        if (!(packetWriter instanceof Channel)) {
-            try {
-                packetWriter.close();
-            } catch (IOException e) {
-                error("preClose({}) Failed ({}) to pre-close packet writer: {}",
-                        this, e.getClass().getSimpleName(), e.getMessage(), e);
+        synchronized (writeState) {
+            writeState.openState = state.get();
+        }
+        super.preClose();
+    }
+
+    @Override
+    protected void doCloseImmediately() {
+        try {
+            // Can't close this in preClose(); a graceful close waits for the currently pending write to finish and thus
+            // still needs the packet writer.
+            if (!(packetWriter instanceof Channel)) {
+                try {
+                    packetWriter.close();
+                } catch (IOException e) {
+                    error("preClose({}) Failed ({}) to pre-close packet writer: {}",
+                            this, e.getClass().getSimpleName(), e.getMessage(), e);
+                }
             }
+            super.doCloseImmediately();
+        } finally {
+            shutdown();
         }
+    }
 
-        super.preClose();
+    private void shutdown() {
+        IoWriteFutureImpl current = null;
+        synchronized (writeState) {
+            writeState.openState = State.Closed;
+            current = writeState.pendingWrite;
+            writeState.pendingWrite = null;
+            writeState.waitingOnIo = false;
+        }
+        if (current != null) {
+            terminateFuture(current);
+        }
+    }
+
+    private void terminateFuture(IoWriteFutureImpl future) {
```

Comment:
Let's make this *protected* in the spirit of allowing users to overload anything they want in our code.

Reply:
Done.